bin/mympirun.py

```diff
@@ -31,92 +31,103 @@
 v3 refactored python 28/08/2012
 v4 cleanup 5/11/2013
 
-Expert mode:
-    export MYMPIRUN_MAIN_EXCEPTION=1 to show all exceptions
-
 TODO:
     intel tuning code
 
-@author: Stijn De Weirdt, Jens Timmerman (HPC UGent / VSC)
+@author: Stijn De Weirdt
+@author: Jens Timmerman
+@author: Jeroen De Clerck
 """
 
-import sys
 import os
+import sys
+import traceback
 
-from vsc.utils import fancylogger
-from vsc.mympirun.mpi.factory import getinstance
-from vsc.mympirun.mpi.mpi import whatMPI
+import vsc.mympirun.mpi.mpi as mpim
 from vsc.mympirun.option import MympirunOption
-from vsc.mympirun.rm.sched import whatSched
+from vsc.mympirun.factory import getinstance
+import vsc.mympirun.rm.sched as schedm
+from vsc.utils import fancylogger
 
 _logger = fancylogger.getLogger()
+fancylogger.setLogLevelDebug()
 
+def get_mpi_and_sched_and_options():
+    """
+    selects mpi flavor and scheduler based on environment and arguments
 
-class ExitException(Exception):
-    """Exception thrown when we wish to exit, but no real errors occured"""
-    pass
-
+    @return: A triplet containing the chosen mpi flavor, chosen scheduler and the MympirunOption class.
+    """
 
-def get_mpi_and_sched_and_options():
-    """Parses the mpi and scheduler based on current environment and guesses the best one to use"""
-    scriptname, mpi, found_mpi = whatMPI(sys.argv[0])
+    scriptname = os.path.basename(os.path.abspath(sys.argv[0]))
+    isfake = scriptname == 'mpirun'
 
-    ismpirun = scriptname == 'mpirun'
+    # init generaloption with the various mpirun cli options
+    mo = MympirunOption(ismpirun=isfake)
 
-    mo = MympirunOption(ismpirun=ismpirun)
+    # see if an mpi flaver was explicitly chosen as an argument
+    # if not, just use the mpirun that was called
+    # We are using sys.argv because generaloption depends on the the returned
+    # scriptname
+    setmpi = mo.options.setmpi if mo.options.setmpi else sys.argv[0]
+    _logger.debug("User chose MPI flavor: %s" % setmpi)
+    scriptname, mpi, found_mpi = mpim.whatMPI(setmpi)
 
     if mo.args is None or len(mo.args) == 0:
         mo.parser.print_shorthelp()
-        raise ExitException("Exit no args provided")
+        _logger.warn("no arguments provided, exiting")
+        sys.exit(0)
 
-    sched, found_sched = whatSched(getattr(mo.options, 'schedtype', None))
+    # Select a Scheduler from the available schedulers
+    sched, found_sched = schedm.whatSched(getattr(mo.options, 'setsched', None))
 
     found_mpi_names = [x.__name__ for x in found_mpi]
     found_sched_names = [x.__name__ for x in found_sched]
 
     if mo.options.showmpi:
         fancylogger.setLogLevelInfo()
-        _logger.info("Found MPI classes %s" % (", ".join(found_mpi_names)))
-        raise ExitException("Exit from showmpi")
+        _logger.info("Found MPI classes %s" %
```

Comment:
one remark for the rest of the PR: the logger does required templating, you can do 
```python
_logger.info("Found MPI classes %s", ", ".join(found_mpi_names))
```